### DISCUSSION TITLE
How convert queryItems into struct

### DISCUSSION THREAD

**Discussion by amine2233** [pointfreeco/swift-url-routing#83] [Q&A]
Hello,

Thank you for your awesome work.

I try to convert an url `"host/hooks/add/repository?name=new-hook&isEnabled&repository=random&events=commit,push,merge"` into a Route but i didn't find an example to how achive this.

my goal is to transform `"name=new-hook&isEnabled&repository=random&events=commit,push,merge"` into `HookCreationDataModel` structure

```swift
import URLRouting
import Parsing

struct AppDeepLinkURLRules {
	let inventoryRouter = OneOf {
		Route(.case(HookTabViewModel.Route.none))
		Route(.case(HookTabViewModel.Route.add(data:route:))) {
			Path { "add" }
			Query {
			    // example ?
			}
		}
	}

	let appRouter = OneOf {
		// Get /notifications
		Route(.case(AppViewModel.Route.Tab.notifications)) {
			Path { "notifications" }
		}

		// Get /hooks
		Route(.case(AppViewModel.Route.Tab.hooks)) {
			Path { "hooks" }
			inventoryRouter
		}

		// Get /settings
		Route(.case(AppViewModel.Route.Tab.settings)) {
			Path { "settings" }
		}
	}
}
```

```swift
final class HookTabRowViewModel: Identifiable, ObservableObject {

	enum Route: Equatable {
		case deleteAlert
		case edit(HookConfigResponse.ID, data: HookCreationDataModel, route: HookEditRoute?)
		case failure(ErrorMessage)
		case toggle(HookConfigResponse.ID)
	}
}

struct HookCreationDataModel: Equatable {
	var name: String = ""
	var isEnabled: Bool = true
	var repository: RepositoriesRowViewModel?
	var tags: [EventRowViewModel] = []
}

final class HookTabViewModel: ObservableObject {

	enum Route: Equatable {
		case none
		case failure(ErrorMessage)
		case row(id: HookTabRowViewModel.ID, route: HookTabRowViewModel.Route?)
		case add(data: HookCreationDataModel, route: HookEditRoute?)
	}
}

final class AppViewModel: ObservableObject {
	enum Route: Equatable {
		case notifications
		case hooks(route: HookTabViewModel.Route?)
		case settings(route: SettingTabViewModels.Route?)

		enum Tab: Equatable {
			case notifications
			case hooks
			case settings
		}

		var selectedTab: Tab {
			switch self {
			case .notifications: return Tab.notifications
			case .hooks: return Tab.hooks
			case .settings: return Tab.settings
			}
		}
	}
}
```

**Comment by stephencelis** [comment]
Hey @amine2233! I don't have time to dig too deeply into this, but a few things stand out that I'm not quite sure of:

- You have `RepositoriesRowViewModel` and `EventRowViewModel` specified inside `HookCreationDataModel`, but you haven't defined them in your post, so there's no way for us to know what data we are parsing into.
- These models sound like observable object classes, which are generally not appropriate to parse into. Instead you most likely want to parse into `struct` and `enum` data types that can be passed to a live view model as configuration.

Hopefully this gets you a bit further. If you have specific `Query` parsing code that you expect to compile but isn't, it might be helpful to share this.
